net/peer_tracker.go

```diff
@@ -0,0 +1,67 @@
+package net
+
+import (
+	"sync"
+
+	logging "github.com/ipfs/go-log"
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+
+	"github.com/filecoin-project/go-filecoin/types"
+)
+
+var logPeerTracker = logging.Logger("peer-tracker")
+
+// PeerTracker is used to record a subset of peers. Its methods are thread safe.
+// It is designed to plug directly into libp2p disconnect notifications to
+// automatically register dropped connections.
+type PeerTracker struct {
+	// mu protects peers
+	mu sync.RWMutex
+	// peers maps stringified peer.IDs to info about their chains
+	peers map[string]*types.ChainInfo
+}
+
+// NewPeerTracker creates a peer tracker.
+func NewPeerTracker() *PeerTracker {
+	return &PeerTracker{
+		peers: make(map[string]*types.ChainInfo),
+	}
+}
+
+// Track adds information about a given peer.ID
+func (tracker *PeerTracker) Track(ci *types.ChainInfo) {
+	tracker.mu.Lock()
+	defer tracker.mu.Unlock()
+
+	pidKey := peer.IDB58Encode(ci.Peer)
```

Comment:
any reason we can't use `ci.Peer.ShortString()`?

Reply:
Just trying to avoid collisions, are the guarantees for `ShortString()` pretty good?

Reply:
Ehh good point (I was thinking of `KeyString` from go-cid), you could just call `ci.Peer.String()`, it's the same `IDB58Encode` call under the hood, just more familiar imo -- your call.

Reply:
Fun fact I started off using `String()` and `IDFromString()` and got burned for over an hour because it turns out those two are not inverses.